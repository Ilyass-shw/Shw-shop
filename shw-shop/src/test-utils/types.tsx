import { configureStore, createSlice } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';

export type StoreType = ReturnType<typeof configureStore>;
export type RenderReturn = ReturnType<typeof render>;
export type SliceParameters = Parameters<typeof createSlice>;
export type UiType = React.ReactElement;

export interface WrapperParams {
  MyStore: StoreType;
  WithRouter?: boolean;
}

export interface useFormProps {
  email: string;
  phone: number;
  buyerAcceptMarketing: string;
}
