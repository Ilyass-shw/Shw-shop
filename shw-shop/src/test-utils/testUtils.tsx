import React, { ReactElement } from 'react';
import { render as rtlRender, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import store, { makeCustomTestStore } from './faksestore';
import { RenderReturn, StoreType, useFormProps } from './types';
import { BrowserRouter } from 'react-router-dom';
import { realStoreType } from '../App/store';
import { useForm, FormProvider } from 'react-hook-form';

export const render = (ui: React.ReactElement): RenderReturn => {
  return rtlRender(ui);
};

// ============= ============= ============= ============= =============

export const renderWithStore = (
  ui: React.ReactElement,
  MyStore: realStoreType,
  WithRouter?: 'withRouter',
): RenderReturn => {
  if (WithRouter) {
    return rtlRender(
      <Provider store={MyStore}>
        <BrowserRouter>{ui}</BrowserRouter>
      </Provider>,
    );
  } else {
    return rtlRender(<Provider store={MyStore}>{ui}</Provider>);
  }
};

// ============= ============= ============= ============= =============

export const renderWithRouter = (ui: React.ReactElement): RenderReturn => {
  return rtlRender(<BrowserRouter>{ui}</BrowserRouter>);
};

// ============= ============= ============= ============= =============

export const makeTestStore = (
  Name?: string,
  initialState?: Record<string, unknown>,
  reducers = {},
): StoreType => {
  if (Name && initialState) {
    const CustomTestStore = makeCustomTestStore(Name, initialState, reducers);
    const origDispatch = CustomTestStore.dispatch;
    CustomTestStore.dispatch = jest.fn(origDispatch);
    return CustomTestStore;
  } else {
    const origDispatch = store.dispatch as jest.Mock;
    store.dispatch = jest.fn(origDispatch);
    return store;
  }
};

// re-export everything
export * from '@testing-library/react';

// ============= ============= ============= ============= =============

export const renderWithFormProvider = (
  ui: ReactElement,
  WithRouter?: 'withRouter',
  { defaultValues = {} } = {},
): RenderResult => {
  // let reactHookFormMethods = {};

  const Wrapper: React.FC = ({ children }) => {
    const methods = useForm({ defaultValues });
    return <FormProvider {...methods}>{children}</FormProvider>;
  };
  const WithRouterWrapper: React.FC = ({ children }) => {
    const methods = useForm({ defaultValues });
    return (
      <BrowserRouter>
        <FormProvider {...methods}>{children}</FormProvider>
      </BrowserRouter>
    );
  };

  if (!WithRouter) {
    return rtlRender(ui, { wrapper: Wrapper });
  }
  return rtlRender(ui, { wrapper: WithRouterWrapper });
};

// ============= ============= ============= ============= =============

export const renderWithReactHookForm = (
  ui: JSX.Element,
  saveData: jest.Mock,
): RenderResult => {
  const Wrapper: React.FC = ({ children }) => {
    const methods = useForm<useFormProps>();
    return (
      <FormProvider {...methods}>
        <form
          onSubmit={methods.handleSubmit((data) => {
            saveData(data);
          })}
        >
          <BrowserRouter>{children}</BrowserRouter>
          <button type="submit" data-testid="testSubmitButton"></button>
        </form>
      </FormProvider>
    );
  };
  return rtlRender(<Wrapper>{ui}</Wrapper>);
};
