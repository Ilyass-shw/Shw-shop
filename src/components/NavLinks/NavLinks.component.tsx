import styled from 'styled-components';
import { RouterLink } from '../../App/App.component';

export const Links = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  padding-right: 0.3rem;
`;

// ============= ============= ============= ============= =============

export const NavLink = styled.div`
  margin: 0 1rem 0 1rem;
`;

// ============= ============= ============= ============= =============

export const LinkButton = styled(RouterLink)`
  font-weight: 700;
  color: black;
  padding-bottom: 1rem;

  @media screen and (min-width: 780px) {
    font-size: 1.2rem;
  }
`;
