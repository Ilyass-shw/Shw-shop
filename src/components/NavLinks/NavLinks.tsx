import React, { useState } from 'react';
import HoverBarWrapper from '../CustomComponents/HoverBarWrapper/HoverBarWrapper/HoverBarWrapper';
import {
  NavLink,
  Links,
  LinkButton,
  CategoryList,
  Category,
} from './NavLinks.component';

const NavLinks: React.FC = () => {
  const [shopHoveredOn, setShopHoveredOn] = useState(false);
  const [caterogiesHoveredOn, setCaterogiesHoveredOn] = useState(false);

  return (
    <Links>
      <NavLink>
        <HoverBarWrapper
          durationInSec={'1s'}
          repeate={1}
          active={shopHoveredOn}
        >
          <LinkButton
            onMouseEnter={(): void => setShopHoveredOn(true)}
            onMouseLeave={(): void => setShopHoveredOn(false)}
            to="/"
          >
            Shop
          </LinkButton>
        </HoverBarWrapper>
      </NavLink>

      <NavLink>
        <HoverBarWrapper
          durationInSec={'1s'}
          repeate={1}
          active={caterogiesHoveredOn}
        >
          <LinkButton
            onMouseEnter={(): void => setCaterogiesHoveredOn(true)}
            onMouseLeave={(): void => setCaterogiesHoveredOn(false)}
            to="/"
          >
            Categories
          </LinkButton>
        </HoverBarWrapper>
        <CategoryList
          data-testid="dropdown Categories"
          onMouseEnter={(): void => setCaterogiesHoveredOn(true)}
          onMouseLeave={(): void => setCaterogiesHoveredOn(false)}
          isHover={caterogiesHoveredOn}
        >
          <Category to="/">Jewelry</Category>
          <Category to="/">Photo prints</Category>
        </CategoryList>
      </NavLink>
    </Links>
  );
};

export default NavLinks;
