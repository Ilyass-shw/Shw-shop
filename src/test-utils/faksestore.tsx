import { configureStore } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import { StoreType } from './types';

const fakeSlice = createSlice({
  name: 'name',
  initialState: {},
  reducers: {},
});

export default configureStore({
  reducer: {
    posts: fakeSlice.reducer,
  },
});

export const makeCustomTestStore = (
  Name: string,
  initialState: Record<string, unknown>,
  reducers = {},
): StoreType => {
  const FakeSlice = createSlice({
    name: Name,
    initialState,
    reducers,
  });

  return configureStore({
    reducer: {
      posts: FakeSlice.reducer,
    },
  });
};
