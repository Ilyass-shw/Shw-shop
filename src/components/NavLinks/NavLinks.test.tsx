import React from 'react';
import { renderWithRouter } from '../../test-utils/testUtils';
import NavLinks from './NavLinks';

describe('NavLinks', () => {
  it('should render', () => {
    renderWithRouter(<NavLinks />);
  });
  it('should match snapshot', () => {
    const { asFragment } = renderWithRouter(<NavLinks />);
    expect(asFragment).toMatchSnapshot();
  });
  it('should match snapshot', () => {
    const { asFragment } = renderWithRouter(<NavLinks />);

    expect(asFragment()).toMatchSnapshot();
  });
});
