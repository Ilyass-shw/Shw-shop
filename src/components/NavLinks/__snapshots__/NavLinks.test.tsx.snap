// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavLinks should match snapshot 1`] = `[Function]`;

exports[`NavLinks should match snapshot 2`] = `
<DocumentFragment>
  <div
    class="sc-eCImvq fFIQqo"
  >
    <div
      class="sc-jRQAMF gwbRxe"
    >
      <div>
        <a
          class="sc-dkPtyc sc-gKckTs cDgGZs gaGKFK"
          href="/"
        >
          Shop
        </a>
        <div
          class="sc-bdvvaa eoYhAo"
        >
          <div
            class="sc-gsDJrp iwdQXL"
          />
        </div>
      </div>
    </div>
    <div
      class="sc-jRQAMF gwbRxe"
    >
      <div>
        <a
          class="sc-dkPtyc sc-gKckTs cDgGZs gaGKFK"
          href="/"
        >
          Categories
        </a>
        <div
          class="sc-bdvvaa eoYhAo"
        >
          <div
            class="sc-gsDJrp iwdQXL"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
